### PR TITLE
Improve Tab navigation behavior with radio buttons inside the dialog

### LINKED ISSUE #439: Incorrect Tab navigation behavior with radio buttons inside the dialog
## Are you reporting a feature or a bug?

Bug

## [Check if the issue is already reported](http://docs.ckeditor.com/#!/guide/dev_issues_readme-section-avoid-duplicates)

http://dev.ckeditor.com/ticket/13839

## Provide detailed reproduction steps (if any)

1. Open Creating Captioned Images sample http://sdk.ckeditor.com/samples/captionedimage.html
1. Double click on the image to get 'Image Properties' dialog
1. Use Tab key to navigate inside the radio button group
1. Keep pressing Tab key

### Expected result

Tab key will enter the radio group.

	When Tab or Shift+Tab into a radio group, focus goes to the selected radio button. If none is selected, focus goes to the first radio button if Tab was pressed, or the last radio bottom if Shift+Tab was pressed.

	When focus is on any radio button, Tab or Shift+Tab will exit the radio group. 

Up Arrow and Left Arrow moves focus to the previous radio button in the group, and selects that button. If focus is on the first item, then focus wraps to last item.

Down Arrow and Right Arrow moves focus to the next radio button in the group, and selects that button. If focus is on the last item, then focus wraps to first item.

Control+Arrow moves through the options without updating content or selecting the button.

Space selects the radio button with focus and de-selects other radio buttons in the group. 

http://www.w3.org/TR/wai-aria-practices/#radiobutton

### Actual result

When focus is on any radio button, Tab will not exit the radio group.

### PR DESCRIPTION
**Pull request by KarolDawidziuk**
## What is the purpose of this pull request?

Bug fix

## Does your PR contain necessary tests?

All patches that change the editor code must include tests. You can always read more
on [PR testing](https://ckeditor.com/docs/ckeditor4/latest/guide/dev_contributing_code.html#tests),
[how to set the testing environment](https://ckeditor.com/docs/ckeditor4/latest/guide/dev_tests.html) and
[how to create tests](https://ckeditor.com/docs/ckeditor4/latest/guide/dev_tests.html#creating-your-own-test)
in the official CKEditor documentation.

### This PR contains

- [x] Unit tests
- [x] Manual tests

## Did you follow the CKEditor 4 code style guide?

Your code should follow the guidelines from the [CKEditor 4 code style guide](https://github.com/ckeditor/ckeditor4/blob/major/dev/docs/codestyle.md) which helps keep the entire codebase consistent.

- [x] PR is consistent with the code style guide

## What is the proposed changelog entry for this pull request?

```
* [#439](https://github.com/ckeditor/ckeditor4/issues/439): Fixed incorrect Tab navigation behavior with radio buttons inside the dialog.
```

## What changes did you make?

*Give an overview…*

It's a second approach of handling this issue based on @jacekbogdanski [proposal](https://github.com/ckeditor/ckeditor4/pull/5368#pullrequestreview-1233838829).

Closes #439.

### CODE REVIEW

File: plugins/dialogui/plugin.js [hunk 1]
```diff
@@ -1249,6 +1252,43 @@ CKEDITOR.plugins.add( 'dialogui', {
...
 		CKEDITOR.ui.dialog.radio.prototype = CKEDITOR.tools.extend( new CKEDITOR.ui.dialog.uiElement(), {
+			focus: function() {
+				var me = this.selectParentTab(),
+					children = me._.children,
+					// Focus the first radio button in the group by default.
+					focusTarget = children[ 0 ],
+					dialog = me._.dialog._,
+					currentFocusIndex = dialog.currentFocusIndex,
+					isMovingBackwards = currentFocusIndex > me.focusIndex,
+					isStartingNewLoop = currentFocusIndex === dialog.focusList.length - 1 && me.focusIndex === 0,
```
[the remark this answers is not in the document]
Comment (jacekbogdanski) [reply plugins/dialogui/plugin.js:1263]: It seems like the issue is caused by https://github.com/ckeditor/ckeditor4/blob/92fa2a55093c6adfb751fcc59c8274a17e2e358e/plugins/dialog/plugin.js#L621-L622 which leads to https://dev.ckeditor.com/ticket/13114#comment:4

As most likely that will be an edge case, we can skip that issue, but please make sure to add a manual test with a dialog that includes tabs to make sure everything works properly (the indexes may be set differently in both cases).

**Review by jacekbogdanski — CHANGES_REQUESTED** [review]
It works pretty well!

One thing I miss is a manual test for dialog that includes some tabs, as it looks like this kind of dialog may behave a bit differently than a tabless one. @KarolDawidziuk could you add that manual test?

Could we have more radio buttons in the last radio group? It looks like an important place to verify if moving from other focusable elements backward (like button) works properly: 
image

File: tests/plugins/dialog/focus.js [hunk 2]
```diff
@@ -290,6 +291,197 @@
...
+					radioElementIndex: 2
+				} ) );
+		},
+
+		'test moving the focus from the last radio group item to the next dialog element by pressing TAB key': function() {
+			var bot = this.editorBot;
+
+			if ( CKEDITOR.env.ie ) {
+				assert.ignore();
+			}
```
[the remark this answers is not in the document]
Comment (KarolDawidziuk) [reply tests/plugins/dialog/focus.js:357]: There was probably an issue with checking radio group elements via code. Looks, that focus is not updating by using our API in unit tests. I can't confirm this because I've no way to open the console on the IE browser.

File: plugins/dialogui/plugin.js [hunk 3]
```diff
@@ -1249,6 +1252,43 @@ CKEDITOR.plugins.add( 'dialogui', {
...
+					isUncheckedCurrentRadioGroup = CKEDITOR.tools.array.every( children, function( child ) {
+						return !child.getInputElement().$.checked;
+					} );
```
[the remark this answers is not in the document]
Comment (KarolDawidziuk) [reply plugins/dialogui/plugin.js:1266]: Yes, we can remove this. 👍

File: plugins/dialogui/plugin.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (KarolDawidziuk) [reply plugins/dialogui/plugin.js]: The manual test which included tabs added.

**Comment by KarolDawidziuk** [comment]
Thanks, @jacekbogdanski for the review.

I've applied improvements mentioned in your comment and PR is ready for another check.

File: plugins/dialogui/plugin.js [hunk 4]
```diff
@@ -1249,6 +1252,43 @@ CKEDITOR.plugins.add( 'dialogui', {
 
 		/** @class CKEDITOR.ui.dialog.radio */
 		CKEDITOR.ui.dialog.radio.prototype = CKEDITOR.tools.extend( new CKEDITOR.ui.dialog.uiElement(), {
+			focus: function() {
+				var me = this.selectParentTab(),
```
[the remark this answers is not in the document]
Comment (KarolDawidziuk) [reply plugins/dialogui/plugin.js]: We can also use `this`. Changed.

File: plugins/dialogui/plugin.js [hunk 5]
```diff
@@ -1249,6 +1254,41 @@ CKEDITOR.plugins.add( 'dialogui', {
...
+					isMovingBackwards = currentFocusIndex > this.focusIndex,
+					isStartingNewLoop = currentFocusIndex === dialogInternal.focusList.length - 1 && this.focusIndex === 0;
+
+				// If focus was changed by using SHIFT + TAB key and the previous radio group
+				// does not have any checked element, focus the last one in the current radio group.
+				// The 'isStartingNewLoop' matters, because in the case when the dialog is shown and the
+				// first element is not focused, then, by default the 'currentFocusIndex' is sets as
+				// the last element from the focus list, and it can be treated as moving focus backwards,
+				// causing that the focus to be incorrectly set to the last element of the radio
+				// group, instead of the first one (#439).
```
Comment (jacekbogdanski) [inline plugins/dialogui/plugin.js:1272]:
```suggestion
				// If focus was changed by using SHIFT + TAB key and the previous radio group
				// does not have any checked element, focus on the last one in the current radio group.
				//
				// When the dialog is shown and the first element is not focused,
				// then, by default the 'currentFocusIndex' is set as
				// the last element from the focus list, and it can be treated as moving focus backward,
				// causing the focus to be incorrectly set to the last element of the radio
				// group, instead of the first one (#439).
```

File: plugins/dialogui/plugin.js [hunk 6]
```diff
@@ -1294,7 +1334,13 @@ CKEDITOR.plugins.add( 'dialogui', {
 				}
 				children[ 0 ].getElement().focus();
 			},
-
+			click: function() {
+				var radioGroup = this.selectParentTab();
+				// Update currentFocusIndex after click on the given radio element. Otherwise, click
+				// will move the focus but the focus index will not be updated, causing moving the
+				// focus incorrectly based on the previous focus index while use Tab or Shift + Tab key (#439).
```
Comment (jacekbogdanski) [inline plugins/dialogui/plugin.js:1340]:
```suggestion
				// Update currentFocusIndex after clicking on the given radio element. Otherwise, click
				// will move the focus but the focus index will not be updated, causing moving the
				// focus incorrectly based on the previous focus index while using Tab or Shift + Tab key (#439).
```

**Review by jacekbogdanski — APPROVED** [review]
Looks good 👍🏻 

@Comandeer could you take a quick look if everything is alright from the accessibility point of view?

**Comment by Comandeer** [comment]
Rebased onto the latest `master`.

**Review by Comandeer — APPROVED** [review]
LGTM!
